Neos.ContentRepository.Core/Classes/Feature/SubtreeTagging/SubtreeTagging.php

```diff
@@ -0,0 +1,124 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Neos\ContentRepository\Core\Feature\SubtreeTagging;
+
+/*
+ * This file is part of the Neos.ContentRepository package.
+ *
+ * (c) Contributors of the Neos Project - www.neos.io
+ *
+ * This package is Open Source Software. For the full copyright and license
+ * information, please view the LICENSE file which was distributed with this
+ * source code.
+ */
+
+use Neos\ContentRepository\Core\ContentRepository;
+use Neos\ContentRepository\Core\DimensionSpace;
+use Neos\ContentRepository\Core\EventStore\Events;
+use Neos\ContentRepository\Core\EventStore\EventsToPublish;
+use Neos\ContentRepository\Core\Feature\Common\ConstraintChecks;
+use Neos\ContentRepository\Core\Feature\Common\NodeAggregateEventPublisher;
+use Neos\ContentRepository\Core\Feature\ContentStreamEventStreamName;
+use Neos\ContentRepository\Core\Feature\SubtreeTagging\Command\TagSubtree;
+use Neos\ContentRepository\Core\Feature\SubtreeTagging\Command\UntagSubtreeTag;
+use Neos\ContentRepository\Core\Feature\SubtreeTagging\Event\SubtreeWasTagged;
+use Neos\ContentRepository\Core\Feature\SubtreeTagging\Event\SubtreeWasUntagged;
+use Neos\EventStore\Model\EventStream\ExpectedVersion;
+
+/**
+ * @internal implementation detail of Command Handlers
+ */
+trait SubtreeTagging
+{
+    use ConstraintChecks;
+
+    abstract protected function getInterDimensionalVariationGraph(): DimensionSpace\InterDimensionalVariationGraph;
+
+    private function handleTagSubtree(TagSubtree $command, ContentRepository $contentRepository): EventsToPublish
+    {
+        $this->requireContentStreamToExist($command->contentStreamId, $contentRepository);
+        $this->requireDimensionSpacePointToExist($command->coveredDimensionSpacePoint);
+        $nodeAggregate = $this->requireProjectedNodeAggregate($command->contentStreamId, $command->nodeAggregateId, $contentRepository);
+        $this->requireNodeAggregateToCoverDimensionSpacePoint(
+            $nodeAggregate,
+            $command->coveredDimensionSpacePoint
+        );
+
+// TODO Adjust to SubtreeTags (is already tagged with same tag => throw exception)
+//        if ($nodeAggregate->disablesDimensionSpacePoint($command->coveredDimensionSpacePoint)) {
+//            // already disabled, so we can return a no-operation.
+//            return EventsToPublish::empty();
+//        }
```

Comment:
This still needs to be done and decided whether this should lead to an exception